File: fungal_py/energy_budget.py
"""Energy budget management and resource constraints."""

import math


class EnergyBudget:
    """Hard resource constraint that gates cycle execution."""

    def __init__(self, initial_budget: int = 1000):
        self.current_budget = initial_budget
        self.total_refunded = 0
        self.total_spent = 0

    def set_budget_from_hardware(self, hardware_profile: dict):
        """Scale budget based on hardware capabilities."""
        cpu_cores = hardware_profile.get('cpu_cores', 4)
        memory_mb = hardware_profile.get('memory_mb', 8192)

        # Base budget scaled by: 1 + log(cores)/10 * memory_factor
        core_factor = 1.0 + math.log(cpu_cores) / 10.0 if cpu_cores > 0 else 1.0
        memory_factor = memory_mb / 8192.0

        self.current_budget = int(1000 * core_factor * (0.8 if memory_factor > 2.0 else memory_factor))

    def spend_for_cycle(self, cost: int) -> bool:
        """Attempt to spend energy for a cycle.

        Returns:
            True if budget available and spent; False if budget exhausted (cycle halts).
        """
        if self.current_budget < cost:
            return False
        self.current_budget -= cost
        self.total_spent += cost
        return True

    def refund_outcome(self, success: bool, cost_paid: int):
        """Refund based on outcome.

        Success gets partial refund (40%).
        Failure applies extra penalty (30 units).
        """
        if success:
            refund = int(cost_paid * 0.40)
            self.current_budget += refund
            self.total_refunded += refund
        else:
            penalty = 30
            self.current_budget -= penalty
            self.total_spent += penalty

File: fungal/src/core/energy_budget.cpp
#include "core/energy_budget.hpp"
#include <sstream>
#include <algorithm>
#include <cmath>

namespace fungal::core {

EnergyBudget::EnergyBudget(int initial_budget)
    : initial_budget_(initial_budget), current_budget_(initial_budget) {}

bool EnergyBudget::can_afford_cycle() const {
    return current_budget_ >= CYCLE_BASE_COST;
}

bool EnergyBudget::spend_for_cycle(int amount) {
    if (amount < 0) {
        return false;
    }
    if (current_budget_ < amount) {
        return false;
    }
    current_budget_ -= amount;
    total_spent_ += amount;
    return true;
}

void EnergyBudget::refund_outcome(bool success, int cost_paid) {
    if (cost_paid < 0) {
        cost_paid = 0;
    }
    if (success) {
        // Partial refund on success (system was right, reward it)
        int refund = static_cast<int>(cost_paid * SUCCESS_REFUND_RATIO / 100.0);
        current_budget_ += refund;
        total_refunded_ += refund;
    } else {
        // Extra penalty on failure (system was wrong, cost more)
        int penalty = FAILURE_PENALTY;
        current_budget_ -= penalty;
        total_spent_ += penalty;
        // Keep budget from drifting into nonsense for Stage1 continuity checks.
        // Economic pressure remains: failures still hurt more than successes help.
        if (current_budget_ < 0) {
            current_budget_ = 0;
        }
    }
}

void EnergyBudget::load_state(int current, int initial, int spent, int refunded) {
    initial_budget_ = initial > 0 ? initial : 0;
    current_budget_ = current;
    if (current_budget_ < 0) {
        current_budget_ = 0;
    }
    total_spent_ = spent >= 0 ? spent : 0;
    total_refunded_ = refunded >= 0 ? refunded : 0;
}

void EnergyBudget::set_budget_from_hardware(const HardwareProfile& profile) {
    // Scale initial budget by hardware capacity
    // More cores and memory = higher budget to explore more
    // Scarce hardware = lower budget, more conservative

    int base_budget = 500;

    // Scale by cores: 4 cores = 1x, 8 cores = 1.3x, 16 cores = 1.5x
    double core_multiplier = 1.0 + (std::log(profile.cpu_cores + 1) / 10.0);

    // Scale by memory: <256MB = 0.5x, 4GB+ = 1.5x
    double memory_multiplier = 1.0;
    if (profile.memory_mb < 256) {
        memory_multiplier = 0.5;
    } else if (profile.memory_mb < 1024) {
        memory_multiplier = 0.8;
    } else if (profile.memory_mb >= 4096) {
        memory_multiplier = 1.5;
    }

    int scaled_budget = static_cast<int>(base_budget * core_multiplier * memory_multiplier);
    initial_budget_ = scaled_budget;
    current_budget_ = scaled_budget;
    total_spent_ = 0;
    total_refunded_ = 0;
}

void EnergyBudget::reset() {
    current_budget_ = initial_budget_;
    total_spent_ = 0;
    total_refunded_ = 0;
}

std::string EnergyBudget::status_string() const {
    std::ostringstream oss;
    oss << "Energy: " << current_budget_ << "/" << initial_budget_
        << " (spent: " << total_spent_ << ", refunded: " << total_refunded_ << ")";
    return oss.str();
}

}  // namespace fungal::core
